csrc/scheduler/matmul_utils.cpp

```diff
@@ -239,28 +233,55 @@ std::string isMatmulFusionDefinitionSupported(
       return roles_map_opt.getErrorMsg();
     }
 
-    size_t core_roles_count = 0;
     const auto& roles_map = roles_map_opt.getData();
-    if (roles_map.count(MatmulRole::MMA_INPUT_A) ==
-        MATMUL_CORE_ROLES_EXPECTED_COUNT) {
-      ++core_roles_count;
+    auto entry = roles_map.find(MatmulRole::INPUT_A);
+    std::set<TensorView*> tvs_with_roles;
+
+    if (entry != roles_map.end()) {
+      if (MATMUL_CORE_ROLES_EXPECTED_COUNT == entry->second.size()) {
+        tvs_with_roles.insert(entry->second.begin(), entry->second.end());
+      } else {
+        return "There is more than a single fusion input that can be MMA first input";
+      }
     } else {
       return "No candidate in fusion inputs for MMA first input";
     }
-    if (roles_map.count(MatmulRole::MMA_INPUT_B) ==
-        MATMUL_CORE_ROLES_EXPECTED_COUNT) {
-      ++core_roles_count;
+
+    entry = roles_map.find(MatmulRole::INPUT_B);
+    if (entry != roles_map.end()) {
+      if (MATMUL_CORE_ROLES_EXPECTED_COUNT == entry->second.size()) {
+        tvs_with_roles.insert(entry->second.begin(), entry->second.end());
+      } else {
+        return "There is more than a single fusion input that can be MMA second input";
+      }
     } else {
       return "No candidate in fusion inputs for MMA second input";
     }
-    if (roles_map.count(MatmulRole::MMA_OUTPUT) ==
-        MATMUL_CORE_ROLES_EXPECTED_COUNT) {
-      ++core_roles_count;
+
+    entry = roles_map.find(MatmulRole::OUTPUT_D);
+    if (entry != roles_map.end()) {
+      if (MATMUL_CORE_ROLES_EXPECTED_COUNT == entry->second.size()) {
+        tvs_with_roles.insert(entry->second.begin(), entry->second.end());
+      } else {
+        return "There is more than a single fusion output that can be MMA output";
+      }
     } else {
       return "No candidate in fusion outputs MMA output";
     }
-    if (roles_map.size() != core_roles_count) {
-      return "Detected fusion roles that are not supported";
+
+    // Non-core roles are optional, no requirements for their presence
+    entry = roles_map.find(MatmulRole::INPUT_C);
+    if (entry != roles_map.end()) {
+      if (entry->second.size() > 1) {
```

Comment:
This is an artificial limitation (up to single `C` input), based on the needs it can be lifted.

Reply:
We should revisit this when we start to have end-to-end matmul integration in the frontend. At that time, we will have a better understanding on what kind of fusion we will need to target.

Reply:
Exactly, although it may seem like an overkill, but introduction of handling more than a single tensor per role should help us with controlling what fusions are supported. It is a different question if it makes sense to have to separate gemms within single kernel (e.g. `(A x B) x C`) but lets keep it for later.